### DISCUSSION TITLE
Package Instantiation (`Web3`, Standalone Packages, and `.use` Functionality)

### DISCUSSION THREAD

**Discussion by Spacesai1or** [ethereum/web3.js#4653] [Ideas]
A user should have 3 methods of instantiating/using Web3 packages:

1. Instantiating the `Web3` class which contains all of the Web3 sub packages
    - Similar to what's made available by [1.x's Web3 package](https://github.com/ChainSafe/web3.js/tree/1.x/packages/web3) 
```typescript
import Web3 from 'web3';
const web3 = new Web3('http://127.0.0.1:8545');
// const web3 = new Web3('http://127.0.0.1:8545', { /* my custom Web3Context */ });

(async () => {
    const protocolVersion = await web3.eth.getProtocolVersion();
    const myContract = web3.eth.contract(myContractAbi);
    const myValue = await myContract.methods.myValue();
})();
```
2. Instantiating specific sub packages to be used as standalone instances
```typescript
import Web3Eth from 'web3-eth';
import Web3EthContract from 'web3-eth-contract';

(async () => {
    const web3Eth = new Web3Eth('http://127.0.0.1:8545');
    // const web3Eth = new Web3Eth('http://127.0.0.1:8545', { /* my custom Web3Context */ });
    const protocolVersion = await web3Eth.getProtocolVersion();


    const myContract = new Web3EthContract(myContractAbi);
    // const myContract = new Web3EthContract(myContractAbi, { /* my custom Web3Context */ });
    const myValue = await myContract.methods.myValue();
})();
```
3. Instantiating some sort of `Web3Core` which contains the bare minimum structure for Web3 to function (including [Web3Context](https://github.com/ChainSafe/web3.js/blob/4.x/packages/web3-core/src/web3_context.ts)), which then allows for sub packages to be appended onto the `Web3Core` instance to make use of `Web3Context`, `requestManager`, etc.
```typescript
import Web3Core from 'web3-core';
import Web3Eth from 'web3-eth';
import Web3EthContract from 'web3-eth-contract';

import myContractAbi = './myContractAbi.json';

const web3Core = new Web3Core('http://127.0.0.1:8545', { /* my custom Web3Context */ });
web3Core.use([Web3Eth, Web3EthContract]);
// web3Core.use(Web3Eth);
// web3Core.use(Web3EthContract);

(async () => {
    const protocolVersion = await web3Core.eth.getProtocolVersion();
    const myContract = web3Core.eth.contract(myContractAbi);
    const myValue = await myContract.methods.myValue();
})();
``` 

---

In method #2, each sub package allows for the specification of a custom `Web3Context`. While flexible, if a change is made to one sub package's `Web3Context` (changing a default value, updating the provider, etc.), none of the other sub package's `Web3Context`s will be affected

In method #1 and #3, the `Web3Context` is managed as global state for all the sub packages via the top level package (`Web3` and `Web3Core`), and if changes are made to `Web3Context`, the global state is updated, and every sub package will be able to listen for the value changes which are emitted using Node's event emitter

**Comment by nazarhussain** [comment]
In current architecture option 1 and 2 are supported. For option 3 need some refacing for the context class but it's doable. Will share the idea later.

**Comment by nazarhussain** [comment]
The earlier discussed functional approach was allowing us to create single global context and pass it around the functions. But in current 4.x rewrite as we moved to class based implementation, each package contains and initiate it's own context. So even for #1 and #3 its not true that there is a global state.

**Comment by nazarhussain** [comment]
Initial implementation of the idea https://github.com/ChainSafe/web3.js/pull/4788
